### PR TITLE
[quant][graph] Add a new observer type for dynamic quantization

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35366 [quant][graph] Add pass to insert quant dequant for dynamic quantization
* #35325 [quant][graph] Update dynamic quant tests to use new qconfig
* **#35265 [quant][graph] Add a new observer type for dynamic quantization**
* #35235 [quant][graph] Add _choose_qparams function for graph mode

Summary:
In graph mode we need to observer the activation tensor for dynamic quantization. This observer should behave the same way as the quantization functions called in the dynamic operator.
Currently for qlinear_dynamic we call quant_utils::ChooseQuantizationParams which has its own logic for calculating scale and zero_point.
We mimic those calculations in the new observer.

Test Plan:
python test/test_quantization.py ObserverTest

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20630988](https://our.internmc.facebook.com/intern/diff/D20630988)

### CODE REVIEW

File: torch/quantization/observer.py [hunk 1]
```diff
@@ -439,6 +439,78 @@ def forward(self, x_orig):
...
+            warnings.warn("Must run observer before calling calculate_qparams.\
+                           Returning default scale and zero point.")
+            return torch.tensor([1.0]), torch.tensor([0])
+
+        assert self.min_val <= self.max_val, "min {} should be less than max {}".format(
+            self.min_val, self.max_val
+        )
+
+        if self.dtype == torch.qint8:
+            if self.reduce_range:
```
Comment (supriyar) [inline torch/quantization/observer.py]: @raghuramank100 I didn't change `reduce_range` in this PR as it also part of MinMaxObserver and `_choose_qparams_per_tensor`. I thought we could have a separate change where we update reduce_range everywhere to `qmin` and `qmax`

File: test/test_quantization.py [hunk 2]
```diff
@@ -1336,6 +1337,25 @@ def test_per_tensor_observers(self, qdtype, qscheme, reduce_range):
...
+
+        X, (scale, zero_point, torch_type) = X
+        x = torch.from_numpy(X)
+        obs = MinMaxDynamicQuantObserver(dtype=torch.quint8, reduce_range=reduce_range)
+
+        result = obs(x)
+        qparams = obs.calculate_qparams()
+        ref = torch._choose_qparams_per_tensor(x, reduce_range)
+
+        np.testing.assert_array_almost_equal(ref[0], qparams[0], decimal=4)
```
Comment (raghuramank100) [inline test/test_quantization.py]: Can we make this match exact?

File: torch/quantization/observer.py (hunk 1 above)
Comment (raghuramank100) [reply torch/quantization/observer.py]: I have a PR for that that I need to update at: https://github.com/pytorch/pytorch/pull/33743

File: torch/quantization/observer.py [hunk 3]
```diff
@@ -439,6 +439,78 @@ def forward(self, x_orig):
...
+                qmin, qmax = 0, 127
+            else:
+                qmin, qmax = 0, 255
+
+        max_val, min_val = float(self.max_val), float(self.min_val)
+        # Extend the min_val and max_val to ensure that it contains 0.
+        min_val = min(0.0, min_val)
+        max_val = max(0.0, max_val)
+        scale = (max_val - min_val) / float(qmax - qmin)
+        if scale == 0.0:
```
Comment (raghuramank100) [inline torch/quantization/observer.py]: Do you also need to check for large values of scale to match with the c++ implementation?

**Review by raghuramank100 — APPROVED** [review]

**Pull request closed without merging**